src/ControlSystem/ControlErrors/Size.hpp

```diff
@@ -0,0 +1,298 @@
+  };
+
+  using options = tmpl::list<MaxNumTimesForZeroCrossingPredictor>;
+  static constexpr Options::String help{
+      "Computes the control error for size control. Will also write a "
+      "diagnostics file if the control systems are allowed to write data to "
+      "disk."};
+
+  /*!
+   * \brief Initializes the `intrp::ZeroCrossingPredictor`s with the number of
+   * times passed in (default 3).
+   *
+   * \details The internal `control_system::size::Info::state` is initialized to
+   * `control_system::size::States::Initial`. All zero crossing predictors are
+   * initialized with a minimum number of times 3. The input argument
+   * `max_times` represents the maximum number of times.
+   */
+  Size(const int max_times = 3) {
+    const size_t max_times_size_t = static_cast<size_t>(max_times);
+    info_.state = std::make_unique<size::States::Initial>();
+    char_speed_predictor_ = intrp::ZeroCrossingPredictor{3, max_times_size_t};
+    comoving_char_speed_predictor_ =
+        intrp::ZeroCrossingPredictor{3, max_times_size_t};
+    delta_radius_predictor_ = intrp::ZeroCrossingPredictor{3, max_times_size_t};
+    legend_ = std::vector<std::string>{"Time",
+                                       "ControlError",
+                                       "StateNumber",
+                                       "DiscontinuousChangeHasOccurred",
+                                       "FunctionOfTime",
+                                       "dtFunctionOfTime",
+                                       "HorizonCoef00",
+                                       "dtHorizonCoef00",
+                                       "MinDeltaR",
+                                       "MinRelativeDeltaR",
+                                       "ControlErrorDeltaR",
+                                       "TargetCharSpeed",
+                                       "MinCharSpeed",
+                                       "MinComovingCharSpeed",
+                                       "CharSpeedCrossingTime",
+                                       "ComovingCharSpeedCrossingTime",
+                                       "DeltaRCrossingTime",
+                                       "SuggestedTimescale",
+                                       "DampingTime"};
+    subfile_name_ =
+        "/ControlSystems/Size" + get_output(Horizon) + "/Diagnostics";
+  }
+
+  /// Returns the internal `control_system::size::Info::suggested_time_scale`. A
+  /// std::nullopt means that no timescale is suggested.
+  const std::optional<double>& get_suggested_timescale() const {
+    return info_.suggested_time_scale;
+  }
+
+  /*!
+   * \brief Check if the `control_system::size::control_error` has decided to
+   * switch states.
+   *
+   * \details If it has decided to switch states, reset the stored
+   * `control_system::size::Info` and return true. If it is staying in the same
+   * state, do nothing and return false.
+   */
+  bool discontinuous_change_has_occurred() {
```

Comment:
While it is documented, it is surprising to me that this function modifies the internal state of the class.  It looks like it should just be querying the state.  What is the reason for this behavior?

If this should be happening, test it.

Reply:
So if a discontinuous change has a occurred, 2 things need to happen

1. Stuff inside the internal `info_` needs to be reset
2. Stuff outside the control error (specifically the Averager) also needs to be reset.

I can add a separate function to the control error that resets things and make this function just query the state. Either way, it's actually hard to test this because we never publicly expose the internal `info_` because nothing outside of this class (except maybe a test) needs to know what's happening. I'm not sure I want to add a member function just for testing purposes. What do you think?

Reply:
Actually never-mind. I already expose the two things that are reset (suggested timescale and discontinuous change) so I can test the new reset function.